### PR TITLE
Apply RFC 8264 enforcement only to non-empty strings

### LINKED ISSUE #2068: Empty strings are not valid RFC 8266 Nicknames
[The enforcement rule for the Nickname Profile in RFC 8266](https://www.rfc-editor.org/rfc/rfc8266#section-2.3) expressly forbids empty strings:

> After all of the foregoing rules have been enforced, the entity MUST ensure that the nickname is not zero bytes in length (this is done after enforcing the rules to prevent applications from mistakenly omitting a nickname entirely, because when internationalized strings are accepted a non-empty sequence of characters can result in a zero-length nickname after canonicalization).

This seems to contradict the recommendation for RPs to set [`PublicKeyCredentialUserEntity.displayName`](https://www.w3.org/TR/webauthn-3/#dom-publickeycredentialuserentity-displayname) to the empty string when "no suitable or [human-palatable](https://www.w3.org/TR/webauthn-3/#human-palatability) name is available" while simultaneously recommending RPs and clients enforce the rule:

> [Relying Parties](https://www.w3.org/TR/webauthn-3/#relying-party) SHOULD perform enforcement, as prescribed in Section 2.3 of [[RFC8266]](https://www.w3.org/TR/webauthn-3/#biblio-rfc8266) for the Nickname Profile of the PRECIS FreeformClass [[RFC8264]](https://www.w3.org/TR/webauthn-3/#biblio-rfc8264), when setting [displayName](https://www.w3.org/TR/webauthn-3/#dom-publickeycredentialuserentity-displayname)'s value, or displaying the value to the user.

> [Clients](https://www.w3.org/TR/webauthn-3/#client) SHOULD perform enforcement, as prescribed in Section 2.3 of [[RFC8266]](https://www.w3.org/TR/webauthn-3/#biblio-rfc8266) for the Nickname Profile of the PRECIS FreeformClass [[RFC8264]](https://www.w3.org/TR/webauthn-3/#biblio-rfc8264), on [displayName](https://www.w3.org/TR/webauthn-3/#dom-publickeycredentialuserentity-displayname)'s value prior to displaying the value to the user or including the value as a parameter of the [authenticatorMakeCredential](https://www.w3.org/TR/webauthn-3/#authenticatormakecredential) operation.

Should the spec be changed to state rule enforcement SHOULD only occur when `displayName` is not empty? I was personally bit by this omission in the RP library I am writing.

### PR DESCRIPTION
**Pull request by selfissued**
Fixes #2068 

Cc @zacknewman

***
<a href="https://pr-preview.s3.amazonaws.com/selfissued/webauthn/pull/2073.html" title="Last updated on Jul 17, 2024, 5:25 PM UTC (787585f)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/webauthn/2073/62b069e...selfissued:787585f.html" title="Last updated on Jul 17, 2024, 5:25 PM UTC (787585f)">Diff</a>

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -3278,8 +3278,8 @@ credential.
     :   <dfn>displayName</dfn>
     ::  A [=human palatability|human-palatable=] name for the [=user account=], intended only for
         display. The [=[RP]=] SHOULD let the user choose this, and SHOULD NOT restrict the choice
-        more than necessary. If no suitable or [=human palatability|human-palatable=] name is
-        available, the [=[RP]=] SHOULD set this value to an empty string.
+        more than necessary.
+        Its value MUST NOT be the empty string.
```
Comment (MasterKale) [inline index.bs:3282]: Hmm, this looks like it could be "an" empty string instead:

```suggestion
        Its value MUST NOT be an empty string.
```

**Review by MasterKale — APPROVED** [review]

**Comment by MasterKale** [comment]
I approved as-is, but I keep coming back to the idea that maybe we keep the existing preamble, but change the "SHOULD set to an empty string" guidance to something like "MUST omit the value" instead 🤔

**Comment by selfissued** [comment]
We could do that.

**Review by timcappalli — APPROVED** [review]

**Comment by emlun** [comment]
>maybe we keep the existing preamble, but change the "SHOULD set to an empty string" guidance to something like "MUST omit the value" instead 🤔

That won't work with the dictionary IDL as currently defined, as `displayName` is a required, non-nullable member.

But maybe we could make it optional? I think the only blocker could be compatibility with CTAP2. But it actually looks like CTAP2 would already be compatible with that change. I noticed this in particular in [§6.8.6. Updating user information](https://fidoalliance.org/specs/fido-v2.1-ps-20210615/fido-client-to-authenticator-protocol-v2.1-ps-errata-20220621.html#updateUserInformation):

>Replace the matching credential’s [`PublicKeyCredentialUserEntity`](https://w3c.github.io/webauthn/#dictdef-publickeycredentialuserentity)'s [`name`](https://w3c.github.io/webauthn/#dom-publickeycredentialentity-name), [`displayName`](https://w3c.github.io/webauthn/#dom-publickeycredentialuserentity-displayname) with the passed-in user details. If a field is not present in the passed `user` details, or it is present and empty, remove it from the matching credential’s [`PublicKeyCredentialUserEntity`](https://w3c.github.io/webauthn/#dictdef-publickeycredentialuserentity).

But one problem is that in CTAP2, `PublicKeyCredentialUserEntity` appears in both input (contravariant) and output (covariant) position. Making `displayName` optional is fine in input position, but could break CTAP2 clients if they read out a `PublicKeyCredentialUserEntity` structure for a discoverable credential and expect `displayName` to always be present. But on the other hand there's also this in the definition of the [authenticatorGetAssertion response structure](https://fidoalliance.org/specs/fido-v2.1-ps-20210615/fido-client-to-authenticator-protocol-v2.1-ps-errata-20220621.html#authenticatorgetassertion-response-structure):

>FIDO Devices - [discoverable](https://fidoalliance.org/specs/fido-v2.1-ps-20210615/fido-client-to-authenticator-protocol-v2.1-ps-errata-20220621.html#discoverable) credentials: For discoverable credentials on FIDO devices, at least [`user`](https://w3c.github.io/webauthn/#dictdef-publickeycredentialuserentity) "[`id`](https://w3c.github.io/webauthn/#dom-publickeycredentialuserentity-id)" is mandatory.

So maybe clients are expected to expect that not all member will always be present.

**Review by emlun — APPROVED** [review]
Should we spell out the same requirement for [`name`](https://w3c.github.io/webauthn/#dom-publickeycredentialentity-name) too?

**Comment by selfissued** [comment]
@emlun, would you be willing to create a PR making `displayName` and `name` optional in the IDL?  I'll then update this PR to indicate that these fields MUST be omitted if they have no values.  Thanks.

**Comment by emlun** [comment]
> @emlun, would you be willing to create a PR making `displayName` and `name` optional in the IDL?

We can't trivially declare `name` as `optional` in the WebIDL because it's inherited from [`PublicKeyCredentialEntity`](https://w3c.github.io/webauthn/#dictdef-publickeycredentialentity) and represents both `rp.name` and `user.name`. So we can't declare `user.name` as `optional` without doing the same with `rp.name`. Which perhaps we could, but we haven't decided to do that as of now.

Also, I think it's a bad idea to make both `name` and `displayName` optional. This was discussed at length in [issue #1915](https://github.com/w3c/webauthn/issues/1915#issuecomment-1617744129).

> Also, we can't make it `optional` without leaving a trap for sites which will then break with all clients that haven't been updated with that.

Oh right, this is also a very good point. In fact we had already come to this same conclusion in #2024.

> Merely amending the spec to explicitly state that Nickname enforcement SHOULD be done by both RPs and clients for _non-empty strings_ is fine with me.

This sounds like a reasonable compromise to me. The background for why the PRECIS enforcement was added is in https://github.com/w3c/webauthn/issues/593#issuecomment-369402225 and I think most of that probably still applies, so I don't think we should drop it completely.

On the other hand there is https://github.com/w3c/webauthn/issues/2068#issuecomment-2119839162:

> OpenID Connect also forbids empty strings as claim values in an attempt to preempt some interop problems. There, https://openid.net/specs/openid-connect-core-1_0.html#UserInfoResponse says:
> 
> > If a Claim is not returned, that Claim Name SHOULD be omitted from the JSON object representing the Claims; it SHOULD NOT be present with a null or empty string value.
> 
> We should do likewise.

But I think this probably overrules still:

> If the value can't be empty, what should sites who don't have a value set it to?

If it can't be optional or nullable, then I guess empty string is the second best alternative?

**Comment by Firstyear** [comment]
> If it can't be optional or nullable, then I guess empty string is the second best alternative?

We specifically enforce it can't be an empty string in webauthn-rs because it's potentially confusing vs a null/none type. And given how many providers offer OIDC, it seems far more prudent to say "must be a string excluding empty string". 

Then you have to also consider how many resident key UI's would mishandle the empty string too in a device listing. So I think there are a fair number of potential ways for this to go sideways.

**Review by sbweeden — DISMISSED** [review]
I don't like the idea that we define this parameter with "`MUST NOT` be the empty string". If we do that, its implied that the client should raise an error if the RP sets this value to the empty string, which would be a breaking change to RPs that might be doing that now. I think it would be better just to define what the client behaviour should be when the empty string is provided (which should be done if we think its an error anyway).

**Comment by MasterKale** [comment]
I'm leaning towards tweaking guidance on how to process `displayName` so that it can be empty, but **when** it's not empty it should be validated according to RFC 8266.

**Comment by nsatragno** [comment]
During the call, we decided: Let's state that the RFC only applies for non-empty values. This prevents us from breaking backwards compatibility, removes the contradiction, and keeps the value from specifying how the display name should be chosen.

The issue discussed during the call of compatibility issues due to empty display names is orthogonal, and likely an implementation issue. If we tried to fix it by making it optional, we'd risk the much worse compatibility risk of clients that haven't updated breaking. We'll figure out if there are any spec bugs leading to that offline with @MasterKale.

(this discussion was long, and I might have missed something worth summarizing, feel free to chime in!)

@selfissued said he'd be happy to update this PR to match.

**Review by emlun — APPROVED** [review]

**Review by nsatragno — APPROVED** [review]
Approved, thank you!

**Review by agl — APPROVED** [review]

**Review by akshayku — APPROVED** [review]

**Pull request merged**